File: src/LeetCode/Algorithms/Java/Happy_Number.java
package LeetCode.Algorithms.Java;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Hashtable;

// https://leetcode.com/problems/happy-number/
public class Happy_Number {
	public static void main(String[] args) {
		int n = 19;
		
		System.out.println(isHappy(n));
	}
	
	public static boolean isHappy(int n) {
		HashSet<Integer> hashSet = new HashSet<>();
		
		hashSet.add(n);
		
		while (n != 1) {
			n = computeNumber(n);
			
			if (hashSet.contains(n)) {
				return false;
			} else {
				hashSet.add(n);
			}
		}
		
		return true;
	}
	
	public static int computeNumber(int n) {
		int sum = 0;
		
		while (n > 0) {
			sum += Math.pow(n % 10, 2);
			n /= 10;
		}
		
		return sum;
	}
}

File: src/LeetCode/Algorithms/JavaScript/Happy_Number.js
/**
 * https://leetcode.com/problems/happy-number/
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function(n) {
	/**
	 * @param {Number} n
	 */
	const computeNumber = (n) => {
		let number = 0;
		
		while (n > 0) {
			number += Math.pow(Math.trunc(n % 10), 2);
			n = Math.trunc(n / 10);
		}
		
		return number;
	};
	
	let hashSet = new Set();
	
	hashSet.add(n);
	
	while (n > 1) {
		n = computeNumber(n);
		
		if (hashSet.has(n)) {
			return false;
		} else {
			hashSet.add(n);
		}
	}
	
	return true;
};

let n = 19;

console.log(isHappy(n));